vulcanai2/tests/models/test_cnn.py

```diff
@@ -1,10 +1,13 @@
 """Test all ConvNet capabilities."""
 import numpy as np
 import pytest
+import copy
 
 import torch
 import torch.nn as nn
 from torch.utils.data import DataLoader, Subset, TensorDataset
+from torch.autograd.gradcheck import gradcheck
```

Comment:
You don't seem to be using `gradcheck`?

Reply:
oops yeah I will removed. Grad check isn't compatible with our data types.

Reply:
why not? seems like something that should be used

Reply:
cause our data is formatted as a list. Take a look at here: https://github.com/pytorch/pytorch/blob/331c4b5b4d600049fbd5a6503c9cf5f1009214c5/torch/autograd/gradcheck.py#L205

Reply:
I tried hacking it, but took a while and couldn't get through it sadly

Reply:
looking at that code it shouldn't be an issue since it will just return the list as a tuple

Reply:
yeah but by default it has to be fed as a tuple. Take a look at here as well: https://discuss.pytorch.org/t/interpreting-gradcheck-errors/16239/11

>

Reply:
Here: https://github.com/pytorch/pytorch/blob/331c4b5b4d600049fbd5a6503c9cf5f1009214c5/torch/autograd/gradcheck.py#L152-L158 , if passed as list will get converted into tuple. And if I pass as tuple(list(data)) then it throws error for requires_grad=True (https://github.com/pytorch/pytorch/blob/331c4b5b4d600049fbd5a6503c9cf5f1009214c5/torch/autograd/gradcheck.py#L223)

Reply:
weird. perhaps the state_dict is getting buried unnecessarily?

Reply:
no it didn't reach the point where forward `(model(date)` or `model.forward(date))` is happening, it just failing around the typecheck within the gradcheck

Reply:
it should be working. I'd like you to look a little more into it if there's any major issues since this should work no problem

Reply:
I think you didn't get what my point is, so it checks firstly if the `inputs` is a list or tuple. Either case it converts to `tuple` and then for each item in the` tuple` it checks for `requires_grad=True` our case will definitely fail, especially if the network is a multi-input_network with one of the `input_network` itself is a multi-input (where our data-set will be a `list` of `list`s). So list doesn't have requires_grad.

Reply:
any updates?

Reply:
no, unfortunately I'm still stuck there.

Reply:
it's a list with embedded lists so loop through and recursively go inside embedded lists and apply this check then if it's not a list.

Reply:
This error seems to be common out there, but couldn't able to find a way to get through it. Even the post ⬆️ I shared is related to that and `.double()` supposed to fix it, but it didn't in this case.

Reply:
Im not sure if `conv1D_net.double()` is enough to cast the entire internal network to double as well. perhaps that can be the issue? if this is causing too many issues and it's a number casting issue then perhaps we should remove this check. but try to first thing I suggested first

Reply:
>but try to first thing I suggested first
 if you mean
>so loop through and recursively go inside embedded lists and apply this check then if it's not a list.
that's what this is: 
>Here is the snippet I tried with:
`gradcheck(conv1D_net.double(), (torch.ones([10, 1, 28], requires_grad=True).double(),))`

isn't it?

Reply:
I mean go through each layer in the network and cast it to double precision instead of the BaseNetwork object.

Reply:
@rfratila I think I found the culprit, its failing [here](https://github.com/Aifred-Health/Vulcan2/blob/test_branch/vulcanai2/tests/models/test_cnn.py#L108), could this be anything to do with the order of `activation` and `pool` ?

Reply:
I don't think the order is the problem since that's quite standard. Don't you need to run inp through the sublayer before doing a gradcheck? also you should use a different var name then net while you go through the sublayers of a network.

Reply:
> Don't you need to run inp through the sublayer

ofcourse I do run the `inp` to the layer, please check: https://github.com/Aifred-Health/Vulcan2/blob/test_branch/vulcanai2/tests/models/test_cnn.py#L99

Reply:
yes you run it through but don't you need to do it before you do a gradcheck with that sublayer?

Reply:
no gradcheck already does a forward inside: https://github.com/pytorch/pytorch/blob/331c4b5b4d600049fbd5a6503c9cf5f1009214c5/torch/autograd/gradcheck.py#L228

Reply:
which makes me realize now:
```
assert gradcheck(net._kernel.double(), (inp.double(),))
inp = net._kernel.double()(inp.double())
```
doesn't makes sense as well I think.

Reply:
try doing one layer at a time instead of the sublayers

Reply:
I tried the before, it was breaking that's why I went sublayer level. I'd appreciate if I can get some help. I've been going back and forth on this.

Reply:
I don't think `gradcheck` is the best thing to use here: https://discuss.pytorch.org/t/check-gradient-flow-in-network/15063

Reply:
no wonder I was breaking my head on that lol